[readonly, for upstream, stacked] Add MD5 checksum to ZIP upload

south-mer opened the pull request:
<!-- 
    Welcome, Athenian! Can you do us two quick favors before you submit your PR?
    
    1. Briefly fill out the sections below. It will make it easy for us to review your code
    2. Put "[WIP]" at the beginning of your PR title if you're not ready to have this merged yet (we have a bot that will tell everyone that it's a work in progress)
-->

## What is the problem I am trying to address?

Continuing the GCP storage backend lock removal, this PR aims to prevent regressions in future changes by precomputing the MD5 checksum and passing the checksum when uploading the file.

The computed checksum can be used for any storage provider; however, it is currently only implemented for GCP backend.

## How is the fix applied?

* Further protect against corrupt cache by precomputing the MD5 sum for the ZIP file.
* Modify the `Saver` interface to accept the MD5 sum, and clarify the contract between the Saver implementation and the Saver user regarding the MD5 sum.
* Implement checksum in upload for **GCP** backend only.
    * Other backends have checksum passed to the implementation, but we did not implement passing the checksum to the upload to prevent breaking existing users since we have no way to test those implementations.

## What GitHub issue(s) does this PR fix or close?

<!--
    If it doesn't fix any GitHub Issues, that's ok. Can you please delete the below "Fixes #" line for us? It would help us out a lot. Thanks!

    Your PR might fix one or more GitHub issues. If so, please use the below "Fixes #<issue number>" notation below. If your PR fixes multiple issues, please put multiple lines of "Fixes #<issue number>", one for each issue. If you do that, when this PR is merged, it'll automatically close the issue(s) you reference.
-->

We are facing cache corruption in production.

<!-- 
example: Fixes #123
-->

allan-mercari commented on pkg/storage/gcp/saver.go:
```diff
@@ -94,6 +94,10 @@ func (s *Storage) upload(ctx context.Context, path string, stream io.Reader, che
 		DoesNotExist: true,
 	}).NewWriter(cancelCtx)
 
+	if len(md5) > 0 {
```
Is this idiomatic for nil check?

south-mer committed:
Fix GCP Storage Backend Locking Issue (#2051)

south-mer committed:
fix: move single-flight to stash (#2050)

south-mer committed:
feat: GCP checksum

dependabot[bot] committed:
chore(deps): bump golang.org/x/oauth2 from 0.23.0 to 0.27.0 (#2058)

Bumps [golang.org/x/oauth2](https://github.com/golang/oauth2) from 0.23.0 to 0.27.0.
- [Commits](https://github.com/golang/oauth2/compare/v0.23.0...v0.27.0)

---
updated-dependencies:
- dependency-name: golang.org/x/oauth2
  dependency-version: 0.27.0
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

mrngm committed:
docfix: use `GONOSUMDB` on Athens' home page (#2056)

shantanugadgil committed:
Update shared-team-instance.md (#2057)
